contracts/validators/CredentialAtomicQueryMTPValidator.sol

```diff
@@ -2,29 +2,104 @@
+        uint256 circuitQueryHash;
+        uint256 requestID;
+        uint256 challenge;
+        uint256 gistRoot;
+        uint256 issuerID;
+        uint256 issuerClaimIdenState;
+        uint256 isRevocationChecked;
+        uint256 issuerClaimNonRevState;
+        uint256 timestamp;
     }
 
-    function getChallengeInputIndex() external pure override returns (uint256 index) {
-        return CHALLENGE_INDEX;
+    function initialize(
+        address _verifierContractAddr,
+        address _stateContractAddr
+    ) public override initializer {
+        _setInputToIndex("merklized", 0);
+        _setInputToIndex("userID", 1);
+        _setInputToIndex("circuitQueryHash", 2);
+        _setInputToIndex("requestID", 3);
+        _setInputToIndex("challenge", 4);
+        _setInputToIndex("gistRoot", 5);
+        _setInputToIndex("issuerID", 6);
+        _setInputToIndex("issuerClaimIdenState", 7);
+        _setInputToIndex("isRevocationChecked", 8);
+        _setInputToIndex("issuerClaimNonRevState", 9);
+        _setInputToIndex("timestamp", 10);
+        _supportedCircuitIds = [CIRCUIT_ID];
+        _circuitIdToVerifier[CIRCUIT_ID] = IVerifier(_verifierContractAddr);
+        super.initialize(_verifierContractAddr, _stateContractAddr);
     }
 
-    function _getInputValidationParameters(
+    function verify(
+        uint256[] calldata inputs,
+        uint256[2] calldata a,
+        uint256[2][2] calldata b,
+        uint256[2] calldata c,
+        bytes calldata data
+    ) external view virtual returns (bool) {
+        CredentialAtomicQuery memory credAtomicQuery = abi.decode(data, (CredentialAtomicQuery));
+        IVerifier verifier = _circuitIdToVerifier[credAtomicQuery.circuitIds[0]];
+
+        require(
+            credAtomicQuery.circuitIds.length == 1 && verifier != IVerifier(address(0)),
+            "Invalid circuit ID"
+        );
+
+        // verify that zkp is valid
+        require(verifier.verify(a, b, c, inputs), "Proof is not valid");
+
+        // parse public signals from inputs array
+        PublicSignals memory signals = parsePublicSignals(inputs);
+
+        // check circuitQueryHash
+        require(
+            signals.circuitQueryHash == credAtomicQuery.queryHash,
+            "Query hash does not match the requested one"
+        );
+
+        // TODO: add support for query to specific userID and then verifying it
+
+        _checkMerklized(signals.merklized, credAtomicQuery.claimPathKey);
+        _checkGistRoot(signals.gistRoot);
+        _checkAllowedIssuers(signals.issuerID, credAtomicQuery.allowedIssuers);
+        _checkClaimIssuanceState(signals.issuerID, signals.issuerClaimIdenState);
+        _checkClaimNonRevState(signals.issuerID, signals.issuerClaimNonRevState);
+        _checkProofExpiration(signals.timestamp);
+        _checkIsRevocationChecked(
+            signals.isRevocationChecked,
+            credAtomicQuery.skipClaimRevocationCheck
+        );
+        return (true);
```

Comment:
Is it the best decision that most of the code is repeated in the verify method in MTP and in SIG circuits instead of be one time represented in its parent class?

Reply:
it is Olexander's decision. Because different PublicSignals are used